docs/root/intro/arch_overview/data_sharing_between_filters.rst

```diff
@@ -0,0 +1,94 @@
+.. _arch_overview_data_sharing_between_filters:
+
+Envoy provides the following mechanisms for the transfer of configuration,
+metadata and per-request/connection state to, from and between filters, as
+well as to other core subsystems (e.g., access logging).
+
+
+Static State
+============
+
+Static state is any immutable state specified at configuration load time
+(e.g., through xDS). There are three categories of static state:
+
+Metadata
+--------
+
+Several parts of Envoy configuration (e.g. listeners, routes, clusters )
+contain a :ref:`metadata <envoy_api_msg_core.Metadata>` where arbitrary
+key-value pairs can be encoded. The typical pattern is to use the filter
+names in reverse DNS format as the key and encode filter specific
+configuration metadata in the value. This metadata is immutable and shared
+across all requests/connections. Such config metadata is usually provided
+during bootstrap time or as part of xDS. For example, weighted clusters in
+HTTP routes use the metadata to indicate the labels on the endpoints
+corresponding to the weighted cluster. Another example, the subset load
+balancer uses the metadata from the route entry corresponding to the
+weighted cluster to select appropriate endpoints in a cluster
+
+Typed Metadata
+--------------
+
+:ref:`Metadata <envoy_api_msg_core.Metadata>` as such is untyped. Before
+acting on the metadata, callers typically convert it to a typed class
+object. The cost of conversion becomes non-negligible when performed
+repeatedly (e.g., for each request stream or connection). Typed Metadata
+solves this problem by allowing filters to register a one time conversion
+logic for a specific key. Incoming config metadata (via xDS) is converted
+to class objects at config load time. Filters can then obtain a typed
+variant of the metadata at runtime (per request or connection), thereby
+eliminating the need for filters to repeatedly convert from
+`ProtobufWkt::Struct` to some internal object during request/connection
+processing.
+
+For example, a filter that desires to have a convenience wrapper class over
+an opaque metadata with key `xxx.service.policy` in `ClusterInfo` could
+register a factory `ServicePolicyFactory` that inherits from
+`ClusterTypedMetadataFactory`. The factory translates the `ProtobufWkt::Struct`
+into an instance of `ServicePolicy` class (inherited from
+`FilterState::Object`). When a `Cluster` is created, the associated
+`ServicePolicy` instance will be created and cached. Note that typed
+metadata is not a new source of metadata. It is obtained from metadata that
+is specified as part of the configuration.
+
+
+HTTP Per-Route Filter Configuration
+-----------------------------------
+
+In HTTP routes, :ref:`per_filter_config
+<envoy_api_field_route.VirtualHost.per_filter_config>` is a special case of
+typed metadata where HTTP filters can have virtualhost/route-specific
```

Comment:
I thought we had agreed to leaved per-route filter config as is? If that's the case, then referencing typed metadata here is a bit confusing to the reader.